File: skills/code-architect/references/code-examples/go/observer.go
// Observer: define a subscription mechanism so multiple observers are
// notified automatically when a subject's state changes.
package main

import "fmt"

// Observer reacts to notifications from a subject.
type Observer interface {
	Update(event string)
}

// subject maintains a list of observers and broadcasts to them.
type subject struct {
	observers []Observer
}

func (s *subject) Subscribe(o Observer) {
	s.observers = append(s.observers, o)
}

func (s *subject) Publish(event string) {
	for _, o := range s.observers {
		o.Update(event)
	}
}

// logger is a concrete observer that records what it hears.
type logger struct {
	name string
	seen []string
}

func (l *logger) Update(event string) {
	l.seen = append(l.seen, event)
	fmt.Printf("%s received %q\n", l.name, event)
}

func main() {
	s := &subject{}
	s.Subscribe(&logger{name: "first"})
	s.Subscribe(&logger{name: "second"})

	s.Publish("state-changed")
}

File: skills/code-architect/references/code-examples/php/Observer.php
<?php
// Observer — subscription mechanism to notify subscribers of events.
declare(strict_types=1);

interface Observer
{
    public function update(string $event): void;
}

interface Subject
{
    public function subscribe(Observer $observer): void;
    public function unsubscribe(Observer $observer): void;
    public function publish(string $event): void;
}

final class EventHub implements Subject
{
    /** @var Observer[] */
    private array $observers = [];

    public function subscribe(Observer $observer): void
    {
        $this->observers[spl_object_id($observer)] = $observer;
    }

    public function unsubscribe(Observer $observer): void
    {
        unset($this->observers[spl_object_id($observer)]);
    }

    public function publish(string $event): void
    {
        foreach ($this->observers as $observer) {
            $observer->update($event);
        }
    }
}

final class LoggingObserver implements Observer
{
    public function __construct(private string $name) {}

    public function update(string $event): void
    {
        echo "{$this->name} received: $event" . PHP_EOL;
    }
}

// Demo
$hub = new EventHub();
$a = new LoggingObserver('A');
$b = new LoggingObserver('B');

$hub->subscribe($a);
$hub->subscribe($b);
$hub->publish('order.created');

$hub->unsubscribe($a);
$hub->publish('order.shipped');
